Replace direct usage of plan constants with a set of functions

adamziel opened the pull request:
Related to https://github.com/Automattic/wp-calypso/pull/23075 and 2-year plans in general (p9jf6J-eR-p2).

We are adding a set of plans to make 2-year subscriptions possible. Current codebase assumes that there is just one variant of each plan type (except for jetpack annual/monthly subscriptions). I can see at least the following problems off the top of my head:

1. Comparisons like `planSlug === PLAN_BUSINESS` would have to check also for `PLAN_BUSINESS_2_YEAR`
2. Code fragments like `<Banner plan={PLAN_BUSINESS} />` may need to receive `PLAN_BUSINESS_2_YEAR` if user is subscribed to a 2-year plan

This PR proposes a solution to above and similar issues. I went through all places where plan constants were used directly and sketched this code and it seems to cover all use cases I saw. Instead of using/comparing plans constants directly, we introduce some metadata to `PLANS_LIST` and also a few helper functions to encapsulate that logic.

We can get rid of some checks like `planSlug === PLAN_BUSINESS`  in favour of checking features, like here: https://github.com/Automattic/wp-calypso/pull/23076 - however this is not possible in all (most) cases.

With this PR, you can check like 
```js
planMatches(planSlug,  {type: TYPE_BUSINESS}); // or shortcut: isBusinessPlan( planSlug );
```

If you need to check just for jetpack or wp.com specific plans, you can do
```js
planMatches(planSlug, {type: TYPE_BUSINESS, group: GROUP_WPCOM})
```

Or, let's say you need a business plan with the same subscription term as active user and related to the same site (jetpack/wp) as current user's plan:
```js
findSimilarPlan(currentPlanSlug, {type: TYPE_BUSINESS})
```

Not only this would allow us to remove some boilerplate code from many components, but would also introduce a consistent way to work with plans - currently they are dealt with in a dozens of different ways.

Finally, with this PR we are able to get rid of other constants such as:
```js
export const JETPACK_PLANS = [
export const JETPACK_MONTHLY_PLANS = [
```
And have PLANS_LIST as a single source of truth allowing us to cross-reference anything in any way without other constants containing additional groupings.

Most important changes are in these two files:
* client/lib/plans/constants.js
* client/lib/plans/index.js

**Test plan:**
~~There is none, this is a prototype and I do not intend to merge this in it's current form. Some function names could be changes etc. My purpose here is to discuss if this way of approaching the problem fits everyone, or if we should take this in a different direction.~~

This change touches number of areas in Calypso:

The following components:
* client/components/banner/index.jsx
* client/components/plans/plan-icon/index.jsx
* client/components/seo/preview-upgrade-nudge/index.jsx

The following functions:
* `getPlanClass()`
* `getMonthlyPlanByYearly()`
* `getYearlyPlanByMonthly()`
* `planLevelsMatch()`

The following selectors:
* `siteHasBusinessPlan()`
* `isBusinessPlanUser()`

It doesn't break any unit tests. It would be best to test it with a full run of e2e tests. In order to test that manually, here's a proposed protocol:

1. Check if purchase/upgrade page with plans displays correct prices and looks properly.
1. Check if purchase/upgrade page with plans displays correct details about current plan and upgrade options for monthly users
1. Check if GoogleMyBusinessStatsNudge is displayed on stats page only for business customers
1. Check if ChatBusinessConciergeNotice in contact form is displayed only to business customers
1. Check if `Upgrade to yearly billing` button on billing page leads to a correct outcome for monthly customers
1. Check if upgrade nudges, such as those in Google Analytics section, have correct plan class (free/personal/premium)
1. Check if upgrade nudges specifically in SEO Preview section show correct details and have correct CSS class

nylen commented:
I _think_ these boolean checks like `isBusinessPlan` should be a pretty straightforward operation, but it is not easy to tell at a glance whether they work as intended.  Why not get rid of `__internal__planMatches` and just write each comparison directly?

```
export function isBusinessPlan( planSlug ) {
    return getPlan( planSlug ).getType() === TYPE_BUSINESS;
}
```

What happens if an invalid plan slug is passed?  What happens if a plan object is passed instead of a plan slug?  (These could be handled in `getPlan`).

`findSimilarPlan` may need more complicated logic this way, but I think that's a better level of abstraction than attempting to handle every possible case in a single query method.

Similarly, I think the test cases for the new functions should be written out one at a time.  This is more boilerplate, but it makes it easier to read what's going on and it makes it much easier to tell at a glance where any failures are occurring.

Note: I am largely new to this code, I'm just approaching it from the perspective of readability.  Overall I think the concept of this PR makes a lot of sense but it could be executed in a more straightforward manner.

adamziel commented:
@nylen 

> I think these boolean checks like isBusinessPlan should be a pretty straightforward operation, but it is not easy to tell at a glance whether they work as intended. Why not get rid of __internal__planMatches and just write each comparison directly?

I'll reply to rest tomorrow but let me just address it quickly - here is a non-exhaustive list of different use-cases we need to address:

* Get all business plans for both Jetpack and wp.com
* Get all premium plans of Jetpack
* Get personal plans of wp.com
* Get business plan for the same site (Jetpack/wp.com) as current user but for 2-year term
* Get premium plan for the same site (Jetpack/wp.com) as current user for the same term as current user
* Get the same type of plan as current user have, but with a 2-year term
* ...and so on

So we need a way to filter and cross-reference those plans in many different ways. That's why `isBusinessPlan` is just a shortcut function that calls a more complex piece of code.

a8ci18n commented on client/lib/plans/constants.js:
```diff
@@ -137,9 +140,242 @@ export const FEATURE_CONCIERGE_SETUP = 'concierge-setup-jetpack';
+	getStoreAudience: () => i18n.translate( 'Best for hobbyists' ),
+	getDescription: () =>
+		i18n.translate(
+			'{{strong}}Best for Personal Use:{{/strong}} Boost your' +
```
Hi! I've found a possible matching string that has already been translated **35** times:
`translate( '{{strong}}Best for Personal Use:{{/strong}} Boost your website with a custom domain name, and remove all WordPress.com advertising. Get access to high-quality email and live chat support.' )` *ES Score: 17*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

adamziel commented:
I made this PR a bit simpler and also rebased it on top of master so it only contains minimum necessary set of changes - it will be easier to reason about it now.

I agree about the simplicity, I stripped part of the API and now it's super simple to reason about. The most prominent change now is addition of `group`, `term`, and `type` data to each plan. Here is how it changes the way we work with plans:

## Example 1

```js
'is-upgrade-personal': includes(
	[ PLAN_PERSONAL, PLAN_JETPACK_PERSONAL, PLAN_JETPACK_PERSONAL_MONTHLY ],	+			{ 'is-upgrade-business': isBusinessPlan( plan ) },
	plan	
),
```

becomes

```js
{ 'is-upgrade-premium': isPremiumPlan( plan ) }
```

## Example 2

```js
switch ( planSlug ) {
	case PLAN_JETPACK_PREMIUM:	
		return PLAN_JETPACK_PREMIUM_MONTHLY;	
	case PLAN_JETPACK_BUSINESS:	
		return PLAN_JETPACK_BUSINESS_MONTHLY;	
	case PLAN_JETPACK_PERSONAL:	
		return PLAN_JETPACK_PERSONAL_MONTHLY;	
	default:	
		return '';	
}
```

becomes

```js
return findSimilarPlan( planSlug, { term: TERM_MONTHLY } );
```

the latter returns plan of the same term as current plan, so monthly, annual, or biennial.

## Example 3

The check for WP.com business plans changes from this:

```js
purchases.some( purchase => PLAN_BUSINESS === purchase.productSlug )
```

to this:

```js
return purchases.some( purchase => {
	const plan = getPlan( purchase.productSlug );
	return plan.type === TYPE_BUSINESS && plan.group === GROUP_WPCOM;
} );
```

This is a bit more code than previously though! And we also should handle a case when plan is null (although I'll also discuss it later in this post). So I propose adding one more helper function and doing the following:

```js
	return purchases.some( purchase => planMatches( purchase.productSlug, {
		type: TYPE_BUSINESS,
		group: GROUP_WPCOM
	} );
```

## Concerns:

> What happens if an invalid plan slug is passed? 

Ideally, I'd like to have `getPlan()` throw an error if invalid slug is passed so we have explicit information that something in the code doesn't work right. It isn's used in that many places and I could refactor them to handle that where it's necessary. I can imagine how right now it's possible to spend many hours tracking a subtle error caused by `getPlan()` returning null in some corner case.

> What happens if a plan object is passed instead of a plan slug? (These could be handled in getPlan).

As suggested, getPlan is a good place to deal with that

> There are also a lot of redux selectors that might need to be brought in line with whichever is chosen to be the new approach.

I updated two of them as part of this PR to show how this would look like, also example 3 comes straight from the selector.

> Using the plan constants directly guarantees that you know which you're including (ie. you won't necessarily include a jetpack site in an upsell for a specific feature of a wordpress.com plan) but also means that products might inadvertently be excluded (such as monthly or biannual version of the plans you want).

That's right! The other end of that problem is that right now when doing just about anything to these constants (adding/removing) we have to make sure the change is included in every single place where they are used, and it's very easy to miss some. So even though it may be tricky at a times, I think we are actually reducing error space here.

The only way I can see to mitigate this problem is having all permutations of `getWPPlans()`, `getJetpackPlans()`, `getWPandJetpackPlans()`, `isWPBusinessPlan()`, `isWPorJetpackBusinessPlan()` etc, but I don't really like this idea.

Also, we should write tests to catch these type of mistakes :-)

@markryall  @dzver  @nylen  @Automattic/calypso-core

nylen commented on client/components/plans/plan-icon/index.jsx:
```diff
@@ -41,18 +29,14 @@ export default class PlanIcon extends Component {
-			case PLAN_JETPACK_BUSINESS:
-			case PLAN_JETPACK_BUSINESS_MONTHLY:
+
+			case isBusinessPlan( plan ):
```
Minor style note: I think it is probably a bit clearer to write this as `if`/`else if`/`else`.

adamziel replied:
I find the switch more readable:

<img width="401" alt="zrzut ekranu 2018-03-12 o 11 19 57" src="https://user-images.githubusercontent.com/205419/37278249-6cd982a4-25e7-11e8-98f0-6c07f6de5f56.png">

taggon replied:
How about using only `if` blocks?
```
if ( isPersonalPlan( plan ) ) {
  return this.getIcon( 'personal' );
}
if ( isPremiumPlan( plan ) ) {
  return this.getIcon( 'premium' );
}
if ( isBusinessPlan( plan ) ) {
  return this.getIcon( 'business' );
}

return this.getIcon( 'free' );
```

adamziel replied:
Great idea!

nylen commented on client/lib/plans/test/plan-lookups.js:
```diff
@@ -0,0 +1,89 @@
+		expect( isBusinessPlan( PLAN_JETPACK_BUSINESS ) ).to.be.true;
+		expect( isBusinessPlan( PLAN_JETPACK_BUSINESS_MONTHLY ) ).to.be.true;
+	} );
+	test( 'should return true for all non-business plans', () => {
```
This test title is incorrect, it should say `false`.

adamziel replied:
Good catch! I fixed the name

nylen commented on client/lib/plans/test/plan-lookups.js:
```diff
@@ -0,0 +1,89 @@
+		expect( isPremiumPlan( PLAN_JETPACK_PREMIUM ) ).to.be.true;
+		expect( isPremiumPlan( PLAN_JETPACK_PREMIUM_MONTHLY ) ).to.be.true;
+	} );
+	test( 'should return true for all non-premium plans', () => {
```
This test title is incorrect, it should say `false`.

adamziel replied:
Good catch! I fixed the name

adamziel commented:
@nylen @taggon I replaced `switch` with a sequence of `if`s

taggon commented on client/lib/plans/index.js:
```diff
@@ -214,11 +198,45 @@ export function getYearlyPlanByMonthly( planSlug ) {
+
+export function planMatches( planKey, query = {} ) {
+	const unknownKeys = difference( Object.keys( query ), [ 'type', 'group', 'term' ] );
+	if ( unknownKeys.length ) {
```
I don't think we need to reject the extra keys here. Can we just ignore them? Or, please specify the error message if you think it is really necessary.

taggon commented:
Awesome! This PR will much improve the readability of the code related to plans! I added a couple of comments. The rest looks good.

nylen commented on client/lib/plans/index.js:
```diff
@@ -214,11 +198,45 @@ export function getYearlyPlanByMonthly( planSlug ) {
+		throw new Error();
+	}
+
+	const plan = getPlan( planKey ) || {}; // @TODO: make getPlan() throw an error on failure
```
This TODO should probably be addressed in this PR, or at least tracked in a separate issue so that it will be followed up on.

nylen commented on client/lib/plans/constants.js:
```diff
@@ -140,7 +141,22 @@ export const FEATURE_CONCIERGE_SETUP = 'concierge-setup-jetpack';
+export const TYPE_FREE = 'TYPE_FREE';
+export const TYPE_PERSONAL = 'TYPE_PERSONAL';
+export const TYPE_PREMIUM = 'TYPE_PREMIUM';
+export const TYPE_BUSINESS = 'TYPE_BUSINESS';
```
Should these constants all be prefixed with `PLAN_`?

nylen commented:
This generally looks pretty good to me, the revised code is much easier to reason about.  I'm not comfortable signing off on it though because I'm not very familiar with plans-related code.

adamziel commented:
Closing this one to follow up with a serie of smaller PRs
